### ISSUE TITLE
[BUG] Embedding fusion acc/auc issue

### ISSUE THREAD

**Issue by marvin-Yu** [changqi1/DeepRec#55]
目前定位出两个问题：
1.	我们fusion的op里面没有做查重筛选。
2.	我们fusion的op里，重复的input的输出却不一致。
Fusion 前向gather的输出，标红为重复部分，但是在反向的输出中并没有重复的元素，连相近的结果都不存在，差距都较大。
``` python
INFO:tensorflow:input_layer/sparse_input_layer/input_layer/C10_embedding/C10_embedding_weights/fused_embedding_lookup_sparse/GatherV2 = 
[[ 1.2214626   0.5241986   0.22209705 -0.56519794  0.12571095  0.26446024
   0.3907412   0.00005793 -0.45320386 -0.69033164  0.3531894  -0.1513126
   0.00713284 -1.1315851   0.12203985  0.23935615]
 [-1.5699558   0.63231647 -0.5136872   0.18575184 -0.12131955 -1.4859123
   0.8938838  -0.33873808 -0.24968442 -0.47764817  0.36187503 -0.14567816
  -0.24810648 -1.3606204  -0.08617076  0.4501951 ]
 [ 1.395356   -0.85390687 -0.7608217   1.0669864   1.191038    0.88764894
   0.9451067   0.29302412  1.2512774   0.6840943  -0.20568915 -0.32980326
  -0.42660442  0.54374695 -0.9136276   0.04837677]
 [ 1.395356   -0.85390687 -0.7608217   1.0669864   1.191038    0.88764894
   0.9451067   0.29302412  1.2512774   0.6840943  -0.20568915 -0.32980326
  -0.42660442  0.54374695 -0.9136276   0.04837677]
 [ 0.6147636   0.33874637 -0.7812209   1.2390836   1.8089103  -1.2311537
  -0.43859923  1.3363832  -0.72441924  1.3167928   1.1064852   0.51790696
  -0.24631402  1.2318567   1.4000374  -0.30377945]
 [-0.3499662  -1.789908    0.48219246  0.2007537   0.7334909  -0.01890297
   0.08424582 -0.9799169  -0.35487846  0.17760478  0.7782412   0.01907562
  -0.5430275  -1.0409418  -0.06544966 -0.31106764]
 [ 1.395356   -0.85390687 -0.7608217   1.0669864   1.191038    0.88764894
   0.9451067   0.29302412  1.2512774   0.6840943  -0.20568915 -0.32980326
  -0.42660442  0.54374695 -0.9136276   0.04837677]
 [-0.5124917   0.45528954  0.7462012   0.20852847  1.4730995   0.8039012
  -0.5750134   0.22652298  1.5296302   0.779812    1.460728    0.8999218
   1.5914694   0.8920278  -1.1893805   1.916351  ]]
```

Fusion grad的输入输出
``` python
输出：
INFO:tensorflow:head/gradients/input_layer/sparse_input_layer/input_layer/C10_embedding/C10_embedding_weights/fused_embedding_lookup_sparse/FusedEmbeddingSparsePostLookUp_grad/FusedEmbeddingSparsePostLookUpGrad = 
[[ 0.00861822  0.0023349  -0.00688701  0.00269023 -0.00164793  0.00736784
  -0.01025489 -0.00652598  0.00471746  0.00888411 -0.00231681  0.00083448
  -0.00203576 -0.00289572  0.00719752 -0.00490604]
 [ 0.00153809 -0.00080411 -0.00177121  0.00086962 -0.00095507  0.00141255
  -0.00152518 -0.0010505   0.00122721  0.00121513 -0.00102509 -0.00052917
  -0.00006346 -0.00056932  0.00194405 -0.00034838]
 [ 0.00154482 -0.00019465 -0.00210896  0.00204525 -0.00301571  0.0016045
  -0.00139093 -0.00215399 -0.00047724  0.00222365  0.00055881 -0.00044712
  -0.00082448  0.00155544  0.00257766  0.00031559]
 [-0.00136614 -0.00111764  0.00218892 -0.00170602  0.00054201 -0.00347374
   0.00119696  0.00144338 -0.00078496 -0.00169556  0.00112028 -0.00118931
  -0.00130751 -0.00075804 -0.00326457 -0.0000487 ]
 [-0.00209746  0.001618    0.00076764  0.00073953  0.00029006 -0.00244137
   0.00196408  0.00168557  0.00034245 -0.00137542 -0.00048502 -0.00033666
  -0.00041434  0.00007309 -0.00190399 -0.00012118]
 [ 0.00003232 -0.00172924  0.00824459 -0.00492863  0.00488566 -0.01386313
   0.00917176  0.0094776   0.00208267 -0.01060377  0.00307963 -0.00334385
   0.00790155 -0.00217232 -0.00438204  0.00839924]
 [-0.003769    0.00456759  0.00200083 -0.00209491  0.00360792 -0.00388297
   0.00045974  0.00181912 -0.00040632 -0.00027408  0.0035839   0.00000899
   0.00114274  0.00211995 -0.00300257  0.00140855]
 [-0.00272977  0.00265015  0.00149372 -0.00279601  0.00245979 -0.00382371
   0.00186041 -0.00059908  0.00158117 -0.00263364  0.00022644 -0.00097163
   0.00043103  0.00025143 -0.00277124 -0.00126195]], 
输入：
head/gradients/input_layer/sparse_input_layer/input_layer/C10_embedding/C10_embedding_weights/Reshape_grad/Reshape = 
[[-0.00861822 -0.0023349   0.00688701 -0.00269023  0.00164793 -0.00736784
   0.01025489  0.00652598 -0.00471746 -0.00888411  0.00231681 -0.00083448
   0.00203576  0.00289572 -0.00719752  0.00490604]
 [-0.00153809  0.00080411  0.00177121 -0.00086962  0.00095507 -0.00141255
   0.00152518  0.0010505  -0.00122721 -0.00121513  0.00102509  0.00052917
   0.00006346  0.00056932 -0.00194405  0.00034838]
 [-0.00154482  0.00019465  0.00210896 -0.00204525  0.00301571 -0.0016045
   0.00139093  0.00215399  0.00047724 -0.00222365 -0.00055881  0.00044712
   0.00082448 -0.00155544 -0.00257766 -0.00031559]
 [ 0.00136614  0.00111764 -0.00218892  0.00170602 -0.00054201  0.00347374
  -0.00119696 -0.00144338  0.00078496  0.00169556 -0.00112028  0.00118931
   0.00130751  0.00075804  0.00326457  0.0000487 ]
 [ 0.00209746 -0.001618   -0.00076764 -0.00073953 -0.00029006  0.00244137
  -0.00196408 -0.00168557 -0.00034245  0.00137542  0.00048502  0.00033666
   0.00041434 -0.00007309  0.00190399  0.00012118]
 [-0.00003232  0.00172924 -0.00824459  0.00492863 -0.00488566  0.01386313
  -0.00917176 -0.0094776  -0.00208267  0.01060377 -0.00307963  0.00334385
  -0.00790155  0.00217232  0.00438204 -0.00839924]
 [ 0.003769   -0.00456759 -0.00200083  0.00209491 -0.00360792  0.00388297
  -0.00045974 -0.00181912  0.00040632  0.00027408 -0.0035839  -0.00000899
  -0.00114274 -0.00211995  0.00300257 -0.00140855]
 [ 0.00272977 -0.00265015 -0.00149372  0.00279601 -0.00245979  0.00382371
  -0.00186041  0.00059908 -0.00158117  0.00263364 -0.00022644  0.00097163
  -0.00043103 -0.00025143  0.00277124  0.00126195]], 
input_layer/sparse_input_layer/input_layer/C10_embedding/C10_embedding_weights/fused_embedding_lookup_sparse/GatherV2 = 
[[ 1.2214626   0.5241986   0.22209705 -0.56519794  0.12571095  0.26446024
   0.3907412   0.00005793 -0.45320386 -0.69033164  0.3531894  -0.1513126
   0.00713284 -1.1315851   0.12203985  0.23935615]
 [-1.5699558   0.63231647 -0.5136872   0.18575184 -0.12131955 -1.4859123
   0.8938838  -0.33873808 -0.24968442 -0.47764817  0.36187503 -0.14567816
  -0.24810648 -1.3606204  -0.08617076  0.4501951 ]
 [ 1.395356   -0.85390687 -0.7608217   1.0669864   1.191038    0.88764894
   0.9451067   0.29302412  1.2512774   0.6840943  -0.20568915 -0.32980326
  -0.42660442  0.54374695 -0.9136276   0.04837677]
 [ 1.395356   -0.85390687 -0.7608217   1.0669864   1.191038    0.88764894
   0.9451067   0.29302412  1.2512774   0.6840943  -0.20568915 -0.32980326
  -0.42660442  0.54374695 -0.9136276   0.04837677]
 [ 0.6147636   0.33874637 -0.7812209   1.2390836   1.8089103  -1.2311537
  -0.43859923  1.3363832  -0.72441924  1.3167928   1.1064852   0.51790696
  -0.24631402  1.2318567   1.4000374  -0.30377945]
 [-0.3499662  -1.789908    0.48219246  0.2007537   0.7334909  -0.01890297
   0.08424582 -0.9799169  -0.35487846  0.17760478  0.7782412   0.01907562
  -0.5430275  -1.0409418  -0.06544966 -0.31106764]
 [ 1.395356   -0.85390687 -0.7608217   1.0669864   1.191038    0.88764894
   0.9451067   0.29302412  1.2512774   0.6840943  -0.20568915 -0.32980326
  -0.42660442  0.54374695 -0.9136276   0.04837677]
 [-0.5124917   0.45528954  0.7462012   0.20852847  1.4730995   0.8039012
  -0.5750134   0.22652298  1.5296302   0.779812    1.460728    0.8999218
   1.5914694   0.8920278  -1.1893805   1.916351  ]], 
input_layer/sparse_input_layer/input_layer/C10_embedding/C10_embedding_weights/fused_embedding_lookup_sparse/FusedEmbeddingSparsePreLookUp:0 = 
[2816  903 6681 6681 1309 1777 6681 5311], 
input_layer/sparse_input_layer/input_layer/C10_embedding/C10_embedding_weights/fused_embedding_lookup_sparse/FusedEmbeddingSparsePostLookUp:1 = [1 1 1 1 1 1 1 1], 
input_layer/sparse_input_layer/input_layer/C10_embedding/C10_embedding_weights/fused_embedding_lookup_sparse/FusedEmbeddingSparsePreLookUp:1 = 
[[0 0]
 [1 0]
 [2 0]
 [3 0]
 [4 0]
 [5 0]
 [6 0]
 [7 0]]
```
Unfusion grad的输入输出：
``` python
输出：INFO:tensorflow:head/gradients/input_layer/sparse_input_layer/input_layer/C10_embedding/C10_embedding_weights/embedding_lookup_sparse_grad/SparseSegmentMeanGrad = 
[[-0.00861822 -0.0023349   0.00688701 -0.00269023  0.00164793 -0.00736784
   0.01025489  0.00652598 -0.00471746 -0.00888411  0.00231681 -0.00083448
   0.00203576  0.00289572 -0.00719752  0.00490604]
 [-0.00153809  0.00080411  0.00177121 -0.00086962  0.00095507 -0.00141255
   0.00152518  0.0010505  -0.00122721 -0.00121513  0.00102509  0.00052917
   0.00006346  0.00056932 -0.00194405  0.00034838]
 [ 0.00359032 -0.00325531 -0.00208078  0.00175567 -0.00113423  0.00575221
  -0.00026577 -0.00110852  0.00166852 -0.00025401 -0.00526298  0.00162744
   0.00098925 -0.00291735  0.00368948 -0.00167544]
 [ 0.00209746 -0.001618   -0.00076764 -0.00073953 -0.00029006  0.00244137
  -0.00196408 -0.00168557 -0.00034245  0.00137542  0.00048502  0.00033666
   0.00041434 -0.00007309  0.00190399  0.00012118]
 [-0.00003232  0.00172924 -0.00824459  0.00492863 -0.00488566  0.01386313
  -0.00917176 -0.0094776  -0.00208267  0.01060377 -0.00307963  0.00334385
  -0.00790155  0.00217232  0.00438204 -0.00839924]
 [ 0.00272977 -0.00265015 -0.00149372  0.00279601 -0.00245979  0.00382371
  -0.00186041  0.00059908 -0.00158117  0.00263364 -0.00022644  0.00097163
  -0.00043103 -0.00025143  0.00277124  0.00126195]], 
输入：
head/gradients/input_layer/sparse_input_layer/input_layer/C10_embedding/C10_embedding_weights/embedding_lookup_sparse_grad/strided_slice = 6, 
head/gradients/input_layer/sparse_input_layer/input_layer/C10_embedding/C10_embedding_weights_grad/tuple/control_dependency_1 = 
[[-0.00861822 -0.0023349   0.00688701 -0.00269023  0.00164793 -0.00736784
   0.01025489  0.00652598 -0.00471746 -0.00888411  0.00231681 -0.00083448
   0.00203576  0.00289572 -0.00719752  0.00490604]
 [-0.00153809  0.00080411  0.00177121 -0.00086962  0.00095507 -0.00141255
   0.00152518  0.0010505  -0.00122721 -0.00121513  0.00102509  0.00052917
   0.00006346  0.00056932 -0.00194405  0.00034838]
 [-0.00154482  0.00019465  0.00210896 -0.00204525  0.00301571 -0.0016045
   0.00139093  0.00215399  0.00047724 -0.00222365 -0.00055881  0.00044712
   0.00082448 -0.00155544 -0.00257766 -0.00031559]
 [ 0.00136614  0.00111764 -0.00218892  0.00170602 -0.00054201  0.00347374
  -0.00119696 -0.00144338  0.00078496  0.00169556 -0.00112028  0.00118931
   0.00130751  0.00075804  0.00326457  0.0000487 ]
 [ 0.00209746 -0.001618   -0.00076764 -0.00073953 -0.00029006  0.00244137
  -0.00196408 -0.00168557 -0.00034245  0.00137542  0.00048502  0.00033666
   0.00041434 -0.00007309  0.00190399  0.00012118]
 [-0.00003232  0.00172924 -0.00824459  0.00492863 -0.00488566  0.01386313
  -0.00917176 -0.0094776  -0.00208267  0.01060377 -0.00307963  0.00334385
  -0.00790155  0.00217232  0.00438204 -0.00839924]
 [ 0.003769   -0.00456759 -0.00200083  0.00209491 -0.00360792  0.00388297
  -0.00045974 -0.00181912  0.00040632  0.00027408 -0.0035839  -0.00000899
  -0.00114274 -0.00211995  0.00300257 -0.00140855]
 [ 0.00272977 -0.00265015 -0.00149372  0.00279601 -0.00245979  0.00382371
  -0.00186041  0.00059908 -0.00158117  0.00263364 -0.00022644  0.00097163
  -0.00043103 -0.00025143  0.00277124  0.00126195]], 
input_layer/sparse_input_layer/input_layer/C10_embedding/C10_embedding_weights/embedding_lookup_sparse/UniqueWithCounts:1 = [0 1 2 2 3 4 2 5], 
input_layer/sparse_input_layer/input_layer/C10_embedding/C10_embedding_weights/embedding_lookup_sparse/Cast = [0 1 2 3 4 5 6 7]
```

**Comment by marvin-Yu** [comment]
@Duyi-Wang  can you put the loss information here? if no any problems, I will close this issue and push this patch to upstream.

**Comment by Duyi-Wang** [comment]
Unfusion
```
INFO:tensorflow:loss = 0.8082987, steps = 1
INFO:tensorflow:loss = 2.507282, steps = 2 (0.025 sec)
INFO:tensorflow:loss = 2.312951, steps = 3 (0.022 sec)
INFO:tensorflow:loss = 1.6560202, steps = 4 (0.022 sec)
INFO:tensorflow:loss = 1.5181658, steps = 5 (0.024 sec)
```
Fusion
```
INFO:tensorflow:loss = 0.8082987, steps = 1
INFO:tensorflow:loss = 2.507282, steps = 2 (0.023 sec)
INFO:tensorflow:loss = 2.312951, steps = 3 (0.020 sec)
INFO:tensorflow:loss = 1.6560202, steps = 4 (0.022 sec)
INFO:tensorflow:loss = 1.5181658, steps = 5 (0.023 sec)
```
